Feature/add docker image

ToshY opened the pull request:
#59 

WIP

* Add initial Dockerfile + .dockerignore
* Add documentation for docker compose
* Updated `docker-compose` to V2 variant `docker compose`
* CI through docker compose (untested)
* CI push docker image to dockerhub on release (untested)

bastien70 commented:
Hi ! Thank you for this beautiful PR! I leave @jmsche  the review, he will know better than me.

However, I think it will merge after my #62  which should change the way adapters are handled. Small changes will certainly be expected, at the level of the documentation that you have updated for example, since we will no longer need to use the environment variables of S3, for the reasons described below.

Thank you very much for the contribution! :)

ToshY commented:
Hey @bastien70,

Thanks for the heads-up. I'll adapt the current documentation in the meantime. I currently have the PR mark as draft as I was still testing if the docker image works correctly in existing projects and with local development.

bastien70 commented:
Hi ! Small question, have you or are you going to write tests to check that the docker container you are setting up works and is correctly configured?

And if I understood correctly, you modified the CI so that the tests are launched thanks to Docker, right? In which case, your PR may be merged before mine.

bastien70 commented:
Hello!

To be more specific on what I was asking:

Basically we have our tests (PHPUnit) which will test the entirety of dbsaver (verify that each feature works). Similarly, how are you going to ensure that the dbsaver docker image you created deploys correctly and without issues?

The goal is that when the CI will launch the PhpUnit tests, it can also check each time that your dbsaver docker image is still compliant, that it deploys correctly, that we have not forgotten anything.

How do you handle this? Maybe you have already answered it, but I must admit that Docker is not our specialty at jmsche and myself, so we will not know if what you have coded is good (but I have the feel like you know what you're doing), or if you've written any tests to make sure the dbsaver docker image works perfectly

bastien70 commented:
Hello @ToshY !

Indeed, now that the tests will be launched based on the docker image, you are right, if from the start it does not work, there is a direct problem with the docker image.

Regarding the docker-release, can you tell me more?

dbsaver docker image is created/updated on every merge? I'm not sure I understand this part correctly.

` I could of course add the phpunit, phpcs and phpstan checks as well in there?`

=> It would be safer yes, even if a priori we are already well thanks to the phpunit tests launched from docker

`Also, I would like to point out that I'm also not an "expert" on docker,......`

=> Although you're not an expert on it, it's still great PR! Good game :)

ToshY commented:
Hey @bastien70

>dbsaver docker image is created/updated on every merge? I'm not sure I understand this part correctly.

Maybe I was unclear about that, but I meant on a new _release_ an image will be build and pushed. Currently in the `docker-release.yml` the following can be found:

```yaml
on:
  release:
    types: [published]
...
```

So it would make sense that upon a new release the workflow gets triggered.

ToshY commented:
I also have a question/request regarding the Taskfile. Using the docker compose commands in the CI makes it visually kind of crowded, so is it acceptable to install Taskfile in the CI? It would reduce the run for each step to just a single task command then.
<sub>p.s. I normally use Makefiles so I don't have this dilemma :wink:</sub>

bastien70 commented:
> I also have a question/request regarding the Taskfile. Using the docker compose commands in the CI makes it visually kind of crowded, so is it acceptable to install Taskfile in the CI? It would reduce the run for each step to just a single task command then. p.s. I normally use Makefiles so I don't have this dilemma wink

Yeah no problem! :)

ToshY commented:
Hey @bastien70 

Hope you don't mind that this is taking some time, but I'll give you an update in the meantime:

* Simplified the commands in the CI by using Taskfile with the help of `arduino/setup-task@v1` action.
* Simplified Dockerfile.
* Added a `docker-compose.dev.yaml` for local development. While it should be relatively easy to use, I will write documentation for it so future contributors (or you 😉) can also use it.

Todo:

* Further testing of dev image, and testing of production image.

One of the things that I think is important is that we should start building building a dev image, tag it as `bastien70/dbsaver:dev` and push it to Dockerhub. This way we can speed up the tests in the CI, because then it only needs to pull the image. If we don't create a dev image it needs to build the image for every run, which is significantly slower compared to just downloading the image.

To do that I actually need to know if you already have a Dockerhub account? And if not, could you make one? As you stated that Docker is not your specialty, I'm not sure if you're familiar with building, tagging and pushing images? If not, the commands are relatively simply but I can guide you through it if needed (already added a command in the Taskfile for building the image). 

If I'm confident enough in the current images, I'll let you know, so that we can start making the dev image before actually merging the PR.

ToshY commented:
Hey @bastien70 

> I have a question, what is the caddy-server for?

Well I realized that if I want to keep contributing to this application, it would be easier for me to add a relatively simple webserver, like caddy, to the `docker-compose.dev.yaml`. So I could just use `task docker:up:dev` to up the services, and after having add `dbsaver.local` / `mail.dbsaver.local` to my systems hosts file, I can easily access the application and mail client through those URLs.  

So to answer your question: "For easier setup/access of the application using docker". 

Even though you may not be very comfortable to docker yet, I hope you/jmsche will try it out before we actually merge the PR. I will of course update the documentation before that to try to make clear on how to use it 😉  

> "Will you still be there to manage docker files?"

Yeah I am still interested in contributing, that includes managing the Dockerfile. However, there shouldn't be a need to change the Dockerfile very often unless there's some (security) issues or need for additional (php) extensions or packages. I will probably still need to update the contribution guidelines to make sure that for new features that require certain extensions/packages that the Dockerfile is updated as well.

ToshY commented:
Hey @bastien70 

To give an update, the last couple of days I've been busy on fixing new CI workflows, making/publishing of the images, and testing the images. To ease the CI testing, I've done this in a temporary private repository, so I can test both the workflows as the images without making a mess here. I'm at the point that everything is green in the CI, so I'm almost ready to commit the last changes back here. I will also include Minio in it then (as you mentioned it in #62 )

I hope to commit my last changes by the end of the week so it can get reviewed 👌

bastien70 commented:
Hello @ToshY 

Greate news! :)
Waiting for your last commit :)

ToshY commented:
Hey @bastien70,

I think it's as good as done. I will review it myself once more this weekend before I'll mark it for review for @jmsche.

I'm missing French documentation regarding the docker compose contribution part, which I just cannot write. From my perspective, it also seems double the work to document in two languages, but maybe we should save that discussion for another time.

Currently the image repository is set to `bastien70/dbsaver`, which means that you should create a repository called `dbsaver` on Dockerhub. This is under the assumption that you've made an account with the username `bastien70`. If you've created your account under another username, please let me know so I can update this.
After you created a Dockerhub account, could you update the repo's (action) secrets by adding `DOCKER_USERNAME` and `DOCKER_PASSWORD`? This is needed so the CI can push the new images to your dockerhub repository.

I also took the liberty to incorporate security-check and dependabot workflows. I could create a separate PR for this if you'd like, but I thought it would be just as easy to include it here as I was overhauling the CI anyway.

ToshY commented on tests/Command/MakeUserCommandTest.php:
```diff
@@ -22,7 +22,7 @@ public function testExecute(): void
         $commandTester->execute(['command' => $command->getName()]);
         $output = $commandTester->getDisplay();
-        self::assertStringContainsString('User contact@test.com has been successfully created! You can now log in.', $output);
+        self::assertStringContainsString('User contact@test.com has been successfully created!', $output);
```
This was updated as it failed due to having a newline between `...created! \n You...`. Think it's good enough to just assert it has the first part.

ToshY commented on tests/Controller/Admin/BackupCrudControllerTest.php:
```diff
@@ -34,14 +37,17 @@ public function testEdit(): void
         $this->loginAsUser();
-        self::$client->request('GET', $url);
-        self::assertResponseRedirects();
+        self::$client->request(Request::METHOD_GET, $url);
```
I could not figure out why this was exactly failing, but for some reason it behaved differently in the CI compared to running tests locally, sometimes returning 302 or 403. Same goes for `UserCurdControllerTest`.

ToshY commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,63 @@
+
+  # Only create database and run migrations for non-test environments
+  if [ "$APP_ENV" != "test" ]; then
+		echo "${CONTAINER_PREFIX} Check database connection..."
```
Strange that the indenting is not correct here. I'll autoformat it.

bastien70 commented:
Hello @ToshY ! Thanks for this PR :)

I will check for docker account on Monday.
I have not yet looked in detail at the files you have modified, but suddenly you confirm that even with this PR it will be possible to launch the project locally and contribute without using docker for those who would like to do it "normally" ?
And so you confirm once again that after this PR merged, it should be very rare that we have to edit it in the future?

ToshY commented:
Hey @bastien70,

> I have not yet looked in detail at the files you have modified, but suddenly you confirm that even with this PR it will be possible to launch the project locally and contribute without using docker for those who would like to do it "normally" ?

I have tried to keep the original way of working unaffected, and with that I mean, in case  you and others still want to develop locally without docker, you should still be able to so (although I haven't tested this myself). 

There are however some changes to existing files which could affect local development:

- The `.env` files are updated to use `mysql:3306` instead of `127.0.0.1:3306` in the `DATABASE_URL`. You would need to create a `.env.local` and override the `DATABASE_URL` back to `127.0.0.1:3306` (in case you have a local MySQL instance).
- The contribution task has slightly changed, but because it already required docker compose before this change, it should not affect your current local development.
- The task commands for the current way of working are kept mostly the same, except for replacing `symfony console` by `bin/console`. This also should not effect your current local development.

> And so you confirm once again that after this PR merged, it should be very rare that we have to edit it in the future?

Normally the Dockerfile itself shouldn't have the need for much update unless you decide to add some features that require additional php extensions, packages or patches for high/severe security updates. As an example, let's say you want to add the php [redis](https://github.com/mlocati/docker-php-extension-installer#supported-php-extensions) extension for a new feature, this means the Dockerfile should simply be appended with the `redis` entry at  `RUN install-php-extensions redis`. If you need additional packages, let's say for example FFmpeg, you can add that by just appending it to `apk add ffmpeg` part.

Regarding security, I made the workflow `docker-base.yml`, which will update both `base` and `dev` images weekly, so that you'll get patches and security updates (from PHP 8.1.X alpine) for those images. When you eventually publish a new release in Github, it will run the `docker-release.yml` workflow, which will build and push `base`, `dev` and `prod` images of the application to Dockerhub.

If you have any more questions let me know 😄

bastien70 commented:
Hello @ToshY !

I just created my docker account and added the secrets keys, you should be able to continue now :)
https://hub.docker.com/repository/docker/bastien70/dbsaver

Tell me if you need anything else :D

ToshY commented:
Hey @bastien70 

Keen eye, I totally forgot that 🤦‍♂️ I will start on that README.

ToshY commented:
Hey @bastien70 

Something like [this](https://gist.github.com/ToshY/cc235a2431e0f2a909606e504e3a5116)?

bastien70 commented:
Hello @ToshY yes it seems good :)

jmsche commented on .github/workflows/coding-standards.yml:
```diff
@@ -10,27 +10,20 @@ on:
 
 jobs:
   php-cs-fixer:
-    name: PHP CS Fixer (PHP ${{ matrix.php-version }})
```
I'd rather have PHP checks run by PHP directly, not through Docker

jmsche replied:
Because nothing requires it?
Remember it's a Symfony app, on which you add an optional Docker image. Docker should remain optional, and in this workflow I see no reason at all to start a Docker container.

ToshY replied:
I'm not sure that I understand the relation between the docker image being optional and not using it in this workflow. If the image is available, why not use it? 

And hypothetically, if I'd bundle the phpcs, static analysis and continuous integration workflows, would you be more open to using the docker container then?

jmsche replied:
Why would the workflow run a Docker image instead of just running PHP? 
I would not run it locally using Docker either if I had PHP available on the machine.

ToshY replied:
Because you don't need to install PHP/composer/packages/libraries in the workflow if it's already available in the image. 

Well if you insist on not using the image, I can just revert it to avoid anymore discussions in order to continue.

jmsche replied:
Please, revert :)

jmsche commented on .github/workflows/security-check.yml:
```diff
@@ -0,0 +1,17 @@
+name: Security Check
```
Unrelated to this PR

jmsche commented on .gitignore:
```diff
@@ -1,29 +1,39 @@
 /.env.local
 /.env.local.php
 /.env.*.local
+/.env.dev
+/.env.prod
```
Can you explain this removal?

ToshY replied:
Could you clarify on what you mean with "removal"? Why I added those 2 files to gitignore?

jmsche commented on .gitignore:
```diff
@@ -1,29 +1,39 @@
-/.task
-/public/coverage
-/public/files/backups
+###> IDE ###
+/.idea
+###< IDE ###
```
This shouldn't be in the project but rather in your own Git global config

jmsche commented on .php-cs-fixer.dist.php:
```diff
@@ -31,5 +31,5 @@
     ])
     ->setFinder($finder)
-    ->setCacheFile(__DIR__.'/var/.php_cs.cache')
+    ->setCacheFile(__DIR__.'/.php-cs-fixer.cache')
```
Why?

ToshY replied:
Think forgot to revert this after testing.

jmsche commented on composer.json:
```diff
@@ -114,16 +116,16 @@
         "cs:dry": "php-cs-fixer fix --diff --dry-run --no-interaction",
         "cs:fix": "php-cs-fixer fix",
-        "phpstan": "vendor/bin/phpstan analyse"
+        "phpstan": "vendor/bin/phpstan analyse --memory-limit 256M"
```
What's the reason for this?

ToshY replied:
Trying to run this inside the docker container with PHP standard memory limit of 128MB will fail. Increasing it helps, but I'm not sure if that's the core problem.

jmsche commented on docs/english.md:
```diff
@@ -117,14 +138,14 @@ Open file `[project]/config/packages/vich_uploader.yaml` and replace its content
-            upload_destination: backup_fs
-    metadata:
-        type: attribute
+  db_driver: orm
+  storage: gaufrette
+  mappings:
+    backups:
+      uri_prefix: '%uploads_base_url%'
+      upload_destination: backup_fs
+  metadata:
+    type: attribute
```
Not required

jmsche commented on docs/french.md:
````diff
@@ -179,12 +198,11 @@ Cette application est protégée par une licence MIT : [LICENCE](../LICENSE).
 ```
 
-Note : vous pouvez lancer ces commandes sans Task, regardez le fichier Taskfile.yml pour voir quelles commandes sont exécutées.
+Note : Vous pouvez lancer ces commandes sans Task, regardez le fichier Taskfile.yml pour voir quelles commandes sont exécutées.
````
Please revert uppercase

jmsche commented on tests/Controller/Admin/BackupCrudControllerTest.php:
```diff
@@ -34,14 +37,17 @@ public function testEdit(): void
+        $url = $this->getCrudActionUrl(Action::DELETE, self::USER_ROLE_USER);
 
-        self::$client->request('GET', $url);
+        self::$client->request(Request::METHOD_GET, $url);
```
Shouldn't be done in a few tests only

jmsche commented on tests/Controller/Admin/UserCrudControllerTest.php:
```diff
@@ -117,12 +118,15 @@ public function testDeleteWithSimpleUser(): void
         $url = $this->getCrudActionUrl(Action::DELETE, self::USER_ROLE_ADMIN);
 
-        self::$client->request('GET', $url);
+        self::$client->request(Request::METHOD_GET, $url);
```
Shouldn't be done in a few tests only

jmsche commented on tests/Controller/Admin/BackupCrudControllerTest.php:
```diff
@@ -34,14 +37,17 @@ public function testEdit(): void
+        $responseCode = self::$client->getResponse()->getStatusCode();
+
+        // If not asserting for both 302 and 403, this will either fail locally or in the CI
+        self::assertTrue(\in_array($responseCode, [Response::HTTP_FOUND, Response::HTTP_FORBIDDEN], true));
```
The test is supposed to pass on both local and CI using the same config. Why would this return a 403 in CI and not locally?

ToshY replied:
I haven't figured out what could cause this to occur, but it somehow behaves differently with redirecting or not.

jmsche commented on tests/Controller/Admin/UserCrudControllerTest.php:
```diff
@@ -117,12 +118,15 @@ public function testDeleteWithSimpleUser(): void
+        $responseCode = self::$client->getResponse()->getStatusCode();
+
+        // If not asserting for both 302 and 403, this will either fail locally or in the CI
+        self::assertTrue(\in_array($responseCode, [Response::HTTP_FOUND, Response::HTTP_FORBIDDEN], true));
```
The test is supposed to pass on both local and CI using the same config. Why would this return a 403 in CI and not locally?

jmsche reviewed (changes requested):
Hi, first set of reviews :)

ToshY commented on .env:
```diff
@@ -27,7 +28,11 @@ APP_SECRET=ed57093ff0c0f5c74d15b176f96d1e66
 # IMPORTANT: You MUST configure your server version, either here or in config/packages/doctrine.yaml
 #
 #DATABASE_URL="sqlite:///%kernel.project_dir%/var/dbsaver.db"
-DATABASE_URL="mysql://root:root@127.0.0.1:3306/dbsaver?serverVersion=5.7"
```
I'm not sure, but as the `DATABASE_URL` is part of the application and not specific to docker, I would keep it inside the current dotenv.

ToshY commented on .env:
```diff
@@ -14,8 +14,9 @@
-APP_SECRET=ed57093ff0c0f5c74d15b176f96d1e66
+APP_ENV="dev"
+APP_SECRET="ed57093ff0c0f5c74d15b176f96d1e66"
+TRUSTED_PROXIES="127.0.0.1,REMOTE_ADDR"
```
Could change it to `TRUSTED_PROXIES=""` if preferred.

jmsche replied:
Yes please :)

ToshY commented:
I haven't had much time to rework the last few weeks, so I'll try to this next week. Also realized that there still may be some small issues with the Dockerfile user permissions, so still have to investigate that further. Sorry for the delay in communication.

ToshY commented:
Due to circumstances, I shall no longer continue my work on this. Sorry for any inconvenience this may have caused.
